src/ucp_ctx/ucc_ucp_ctx.c

```diff
@@ -0,0 +1,183 @@
+        ucp_params.field_mask |= UCP_PARAM_FIELD_ESTIMATED_NUM_EPS;
+        ucp_params.estimated_num_eps = params->estimated_num_eps;
+    }
+
+    if (params->mask & UCC_UCP_CTX_PARAM_FIELD_DEVICES) {
+        ucp_config_modify(ucp_config, "NET_DEVICES", params->devices);
+    }
+
+    status = ucp_init(&ucp_params, ucp_config, &ucp_context);
+    ucp_config_release(ucp_config);
+    if (UCS_OK != status) {
+        ucc_error("failed to init ucp context, %s", ucs_status_string(status));
+        ucc_status = ucs_status_to_ucc_status(status);
+        goto err_cfg_read;
+    }
+
+    worker_params.field_mask = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    ucc_assert(params->mask & UCC_UCP_CTX_PARAM_FIELD_THREAD_MODE);
+    switch (params->thread_mode) {
+    case UCC_THREAD_SINGLE:
+    case UCC_THREAD_FUNNELED:
+        worker_params.thread_mode = UCS_THREAD_MODE_SINGLE;
+        break;
+    case UCC_THREAD_MULTIPLE:
+        worker_params.thread_mode = UCS_THREAD_MODE_MULTI;
+        break;
+    default:
+        /* unreachable */
+        ucc_assert(0);
+        break;
+    }
+    status = ucp_worker_create(ucp_context, &worker_params, &ucp_worker);
+    if (UCS_OK != status) {
+        ucc_error("failed to create ucp worker, %s", ucs_status_string(status));
+        ucc_status = ucs_status_to_ucc_status(status);
+        goto err_worker_create;
+    }
+
+    if (params->thread_mode == UCC_THREAD_MULTIPLE) {
+        worker_attr.field_mask = UCP_WORKER_ATTR_FIELD_THREAD_MODE;
+        ucp_worker_query(ucp_worker, &worker_attr);
+        if (worker_attr.thread_mode != UCS_THREAD_MODE_MULTI) {
+            ucc_error("thread mode multiple is not supported by ucp worker");
+            status = UCC_ERR_NOT_SUPPORTED;
+            goto err_thread_mode;
+        }
+    }
+
+    ctx_h->ctx.ucp_context = (void *)ucp_context;
+    ctx_h->ctx.ucp_worker  = (void *)ucp_worker;
+    ctx_h->ref_count       = 0;
+    *ctx                   = ctx_h;
+    return UCC_OK;
+
+err_thread_mode:
+    ucp_worker_destroy(ucp_worker);
+err_worker_create:
+    ucp_cleanup(ucp_context);
+err_cfg_read:
+    free(ctx_h);
+    return ucc_status;
+}
+
+ucc_status_t ucc_ucp_ctx_destroy(ucc_ucp_ctx_handle_t *ctx_h)
+{
+    ucp_context_h ucp_context;
+    ucp_worker_h  ucp_worker;
+    if (ctx_h->ref_count != 0) {
+        ucc_warn(
+            "destroying ucc_ucp_ctx %p which is still in use, ref_count %d",
+            ctx_h, ctx_h->ref_count);
+    }
+    ucp_context = (ucp_context_h)ctx_h->ctx.ucp_context;
+    ucp_worker  = (ucp_worker_h)ctx_h->ctx.ucp_worker;
+
+    ucp_worker_destroy(ucp_worker);
+    ucp_cleanup(ucp_context);
+    free(ctx_h);
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_ucp_ctx_get(ucc_ucp_ctx_handle_t *ctx_handle,
+                             ucc_ucp_ctx_t **ctx, ucc_ucp_ctx_type_t ctx_type)
+{
+    /* protect tag space: don't allow the same ctx_type to be used
+       on the same ucp context */
+    if (ctx_handle->types_used[ctx_type] >= 0) {
```

Comment:
how it can become >=0?

Reply:
if some component grabs the instance of ucc_ucp_ctx_t with a given ucc_ucp_ctx_type_t and then another component will attempt to grab with the same ucc_ucp_ctx_type_t. It could potentially lead to tag collision. It can only happen if there is a devel error. I agree maybe it is too much (i was in doubt when added this). Ref count is enough. I'll remove this.